Add examples of __repr__ for datatypes and add example of type check error

cosmicexplorer opened the pull request:
### Problem

In #5759, which updated the engine readme, there were no examples of what the `repr` of datatypes looked like. There was also a mention of a type check error, but no description of when or how that would occur.

### Solution

Provide examples of `str` and `repr` for a datatype with and without typed fields, as well as an example of a type check error on construction.

stuhood commented:
Sorry, I wasn't clear: I was suggesting that both the `str` and `repr` of a `datatype` should use the `repr` of all fields... otherwise this kind of thing is ambiguous:
```
> class FormattedInt(datatype(['f1', 'f2'])): pass
> print(FormattedInt("a string containing, f3=", "and stuff"))
FormattedInt(f1=a string containing, f3=, f2=and stuff)
```
... ie, punctuation in the fields ends up interspersed in a confusing way with the field names.

cosmicexplorer commented:
That makes sense, thanks for clarifying. I'll make the change, edit the test cases, and update this pr.

cosmicexplorer commented:
Rebased and updated the docs change, but still need to address the `__str__` output issue mentioned above. The only reason we wouldn't want to use the `repr()` of datatype field values is because if the value itself is a datatype, it would then recursively call `repr()` on all of *its* fields, including non-datatype fields. As I'm typing this now I'm realizing (lol) that we can simply make an auxiliary method for printing field values which special-cases datatype objects (which we can identify now after using the `DatatypeMixin` from #7304).

cosmicexplorer commented:
Interesting to consider in the future, but this can be closed now.
